From 6451d1cf979cc449bd67ba6f01861bc7b41dc971 Mon Sep 17 00:00:00 2001
From: Erhan Bilgili <erhan.bilgili@gmail.com>
Date: Thu, 20 Aug 2026 00:04:57 +0200
Subject: [PATCH] win32u: Remove the optional window move hack.

WINE_MOVE_HACK adjusted application window rectangles to keep them
inside the current virtual screen. This rewrote requested Win32 geometry
in the generic window-position path and could conflict with display
driver output placement.

Remove the workaround and its environment variable. Winewayland now
handles output and layout transitions through explicit fullscreen and
compositor state.

(cherry picked from commit 6451d1cf979cc449bd67ba6f01861bc7b41dc971)

Source: https://github.com/nanomatters/wine-salkim/commit/6451d1cf979cc449bd67ba6f01861bc7b41dc971
---
 dlls/win32u/window.c | 64 +-------------------------------------------
 1 file changed, 1 insertion(+), 63 deletions(-)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index bef8841b245..a1a56a524c4 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -2325,19 +2325,6 @@ static BOOL is_fullscreen( const MONITORINFO *info, const RECT *rect )
            rect->top <= info->rcMonitor.top && rect->bottom >= info->rcMonitor.bottom;
 }
 
-static int use_move_hack(void)
-{
-    static volatile int enabled = -1;
-
-    if (enabled == -1)
-    {
-        const char *env = getenv("WINE_MOVE_HACK");
-        enabled = (env && !strcmp(env, "1"));
-    }
-
-    return enabled;
-}
-
 /***********************************************************************
  *           apply_window_pos
  *
@@ -2351,61 +2338,12 @@ static BOOL apply_window_pos( HWND hwnd, HWND insert_after, UINT swp_flags, stru
     HWND owner_hint, surface_win = 0, toplevel;
     UINT raw_dpi, monitor_dpi, dpi = get_thread_dpi();
     BOOL ret, is_layered, is_child, need_icons = FALSE;
-    struct window_rects old_rects, adjusted;
+    struct window_rects old_rects;
     RECT extra_rects[3];
     struct window_surface *old_surface;
     HICON icon, icon_small;
     ICONINFO ii, ii_small;
 
-    /* HACK: move windows within the virtual screen on winewayland */
-    if (use_move_hack() && !(get_window_long( hwnd, GWL_STYLE ) & WS_MAXIMIZE))
-    {
-        RECT temp, *adj;
-        RECT virtual_screen = get_virtual_screen_rect( get_thread_dpi(), MDT_DEFAULT );
-
-        adjusted = *new_rects;
-        adj = &adjusted.client;
-
-        intersect_rect(&temp, &virtual_screen, adj);
-
-        /* we aren't off screen */
-        if (!IsRectEmpty(&temp))
-        {
-            LONG offset_x = 0, offset_y = 0;
-
-            if (adj->bottom - adj->top <=
-                virtual_screen.bottom - virtual_screen.top)
-            {
-                if (adj->bottom > virtual_screen.bottom)
-                    offset_y = virtual_screen.bottom - adj->bottom;
-                else if (virtual_screen.top > adj->top)
-                    offset_y = virtual_screen.top - adj->top;
-            }
-
-            if (adj->right - adj->left <=
-                virtual_screen.right - virtual_screen.left)
-            {
-                if (adj->right > virtual_screen.right)
-                    offset_x = virtual_screen.right - adj->right;
-                else if (virtual_screen.left > adj->left)
-                    offset_x = virtual_screen.left - adj->left;
-            }
-
-            OffsetRect(&adjusted.client, offset_x, offset_y);
-            OffsetRect(&adjusted.visible, offset_x, offset_y);
-            OffsetRect(&adjusted.window, offset_x, offset_y);
-
-            if (offset_x != 0 || offset_y != 0)
-            {
-                TRACE("vscreen rect %s\n", wine_dbgstr_rect(&virtual_screen));
-                TRACE("Original window rects: %s\n", debugstr_window_rects(new_rects));
-                TRACE("Adjusted window rects: %s\n", debugstr_window_rects(&adjusted));
-
-                new_rects = &adjusted;
-            }
-        }
-    }
-
     toplevel = NtUserGetAncestor( hwnd, GA_ROOT );
     is_layered = new_surface && new_surface->alpha_mask;
     is_child = toplevel && toplevel != hwnd;
-- 
2.55.0

